addons/l10n_in_edi/models/account_edi_format.py

```diff
@@ -0,0 +1,573 @@
+            supply_type = move.l10n_in_gst_treatment == "overseas" and "EXPWOP" or "SEZWOP"
+        elif move.l10n_in_gst_treatment == "deemed_export":
+            supply_type = "DEXP"
+        return supply_type
+
+    #================================ API methods ===========================
+
+    @api.model
+    def _l10n_in_edi_no_config_response(self):
+        return {'error': [{
+            'code': '000',
+            'message': _(
+                "A username and password still needs to be set or it's wrong for the E-invoice(IN). "
+                "It needs to be added and verify in the Settings."
+            )}
+        ]}
+
+    @api.model
+    def _l10n_in_edi_get_token(self, company):
+        sudo_company = company.sudo()
+        if sudo_company.l10n_in_edi_username and sudo_company._l10n_in_edi_token_is_valid():
+            return sudo_company.l10n_in_edi_token
+        elif sudo_company.l10n_in_edi_username and sudo_company.l10n_in_edi_password:
+            self._l10n_in_edi_authenticate(company)
+            return sudo_company.l10n_in_edi_token
+        return False
+
+    @api.model
+    def _l10n_in_edi_connect_to_server(self, company, url_path, params):
+        user_token = self.env["iap.account"].get("l10n_in_edi")
+        params.update({
+            "account_token": user_token.account_token,
+            "dbuuid": self.env["ir.config_parameter"].sudo().get_param("database.uuid"),
+            "username": company.sudo().l10n_in_edi_username,
+            "gstin": company.vat,
+        })
+        if company.sudo().l10n_in_edi_production_env:
+            default_endpoint = DEFAULT_IAP_ENDPOINT
+        else:
+            default_endpoint = DEFAULT_IAP_TEST_ENDPOINT
+        endpoint = self.env["ir.config_parameter"].sudo().get_param("l10n_in_edi.endpoint", default_endpoint)
+        url = "%s%s" % (endpoint, url_path)
+        try:
+            return jsonrpc(url, params=params, timeout=25)
+        except AccessError as e:
+            _logger.warning("Connection error: %s", e.args[0])
+            return {
+                "error": [{
+                    "code": "404",
+                    "message": _("Unable to connect to the online E-invoice service."
+                        "The web service may be temporary down. Please try again in a moment.")
+                }]
+            }
+
+    @api.model
+    def _l10n_in_edi_authenticate(self, company):
+        params = {"password": company.sudo().l10n_in_edi_password}
+        response = self._l10n_in_edi_connect_to_server(company, url_path="/iap/l10n_in_edi/1/authenticate", params=params)
+        # validity data-time in Indian standard time(UTC+05:30) so remove that gap and store in odoo
+        if "data" in response:
+            local_time = fields.Datetime.context_timestamp(
+                self.with_context(tz="Asia/Kolkata"),
+                fields.Datetime.to_datetime(response["data"]["TokenExpiry"]
```

Comment:
just to double check `response["data"]["TokenExpiry"]` is in UTC, right? Not in `Asia/Kolkata` timezone?

Reply:
@mart-e No, it's given time in the Indian time zone(Asia/Kolkata)

Reply:
Then, `context_timestamp` is not what you want. See [the docstring](https://github.com/odoo/odoo/blob/35ac2fa584fa7304015a9b44573622cb820d390a/odoo/fields.py#L1971)
```
        :param datetime timestamp: naive datetime value (expressed in UTC)
            to be converted to the client timezone.
        :return: timestamp converted to timezone-aware datetime in context timezone.
```

Reply:
@mart-e yes, right.
If the DateTime string comes in a specific time zone but does not have timezone on that string so as I test we have two solutions plz suggest.

1. using timezone
```
a = fields.Datetime.to_datetime('2022-04-12 00:00:00').replace(tzinfo=pytz.timezone('Asia/Kolkata'))
>>> a.astimezone(pytz.utc)
datetime.datetime(2022, 4, 11, 18, 7, tzinfo=<UTC>)
```

2. directly convert to UTC by - 5:30

```
>>> fields.Datetime.to_datetime('2022-04-12 00:00:00') - timedelta(hours=5, minutes=30, seconds=00)
datetime.datetime(2022, 4, 11, 18, 30)
```

if any other solution then also plz suggest.

Reply:
Ohh ,I see there is +5:53 using time zone instead of +5:30
```datetime.datetime(2022, 4, 12, 0, 0, tzinfo=<DstTzInfo 'Asia/Kolkata' LMT+5:53:00 STD>)```
timezone base is really change!!!!!

Reply:
okay replace give issue so we use localize instead of replace
```a = tz.localize(fields.Datetime.to_datetime('2022-04-12 00:00:00')) 
>>> a
datetime.datetime(2022, 4, 12, 0, 0, tzinfo=<DstTzInfo 'Asia/Kolkata' IST+5:30:00 STD>)
>>> a.astimezone(pytz.utc)
datetime.datetime(2022, 4, 11, 18, 30, tzinfo=<UTC>)```

Reply:
@mart-e this is fixed plz check